crml/cennzx-spot/src/lib.rs

```diff
@@ -657,59 +657,73 @@ impl<T: Trait> Module<T> {
 			return Ok(T::Balance::max_value());
 		}
 
-		let amount = U256::from(T::BalanceToU128::from(output_amount).into());
-		let input_reserve = U256::from(T::BalanceToU128::from(input_reserve).into());
-		let denominator = U256::from(T::BalanceToU128::from(output_reserve - output_amount).into());
-
-		let res: u128 = (input_reserve * amount / denominator)
-			.try_into()
-			.map_err(|_| "Overflow error")?;
-
-		let price = T::U128ToBalance::from(res).into();
-		let price_plus_one = T::BalanceToU128::from(price + One::one());
-		let output = FeeRate::safe_mul(FeeRate::one() + fee_rate, price_plus_one)?;
-		Ok(T::U128ToBalance::from(output).into())
+		let output_amount_hp = HighPrecisionUnsigned::from(T::BalanceToUnsignedInt::from(output_amount).into());
+		let output_reserve_hp = HighPrecisionUnsigned::from(T::BalanceToUnsignedInt::from(output_reserve).into());
+		let input_reserve_hp = HighPrecisionUnsigned::from(T::BalanceToUnsignedInt::from(input_reserve).into());
+		let denominator_hp = output_reserve_hp - output_amount_hp;
+		let price_hp = input_reserve_hp
+			.saturating_mul(output_amount_hp)
+			.checked_div(denominator_hp)
+			.ok_or::<&'static str>(Error::DivideByZero.into())?;
+
+		let price_lp_result: Result<LowPrecisionUnsigned, &'static str> =
+			LowPrecisionUnsigned::try_from(price_hp).map_err(|_| Error::Overflow.into());
+		let price_lp = price_lp_result?;
+		let price_plus_one = price_lp
+			.checked_add(One::one())
+			.ok_or::<&'static str>(Error::Overflow.into())?;
+		let fee_rate_plus_one = fee_rate
+			.checked_add(FeeRate::<PerMill>::one())
+			.ok_or::<&'static str>(Error::Overflow.into())?;
+		let output = fee_rate_plus_one
+			.checked_mul(price_plus_one.into())
+			.ok_or::<&'static str>(Error::Overflow.into())?;
+		Ok(T::UnsignedIntToBalance::from(output.into()).into())
 	}
 
 	fn get_input_price(
 		input_amount: T::Balance,
 		input_reserve: T::Balance,
 		output_reserve: T::Balance,
-		fee_rate: FeeRate,
+		fee_rate: FeeRate<PerMill>,
 	) -> rstd::result::Result<T::Balance, &'static str> {
 		if input_reserve.is_zero() || output_reserve.is_zero() {
-			return Err("Pool is empty");
+			return Err(Error::EmptyPool.into());
 		}
 
-		let div_rate = FeeRate::one() + fee_rate;
-
-		let lhs = T::BalanceToU128::from(input_amount);
+		let div_rate: FeeRate<PerMill> = fee_rate
+			.checked_add(FeeRate::<PerMill>::one())
+			.ok_or::<&'static str>(Error::Overflow.into())?;
 
-		let input_amount_less_fee_scaled = FeeRate::safe_div(lhs, div_rate)?;
-		let input_reserve: u128 = T::BalanceToU128::from(input_reserve).into();
-		let output_reserve = U256::from(T::BalanceToU128::from(output_reserve).into());
-		let input_amount = U256::from(input_amount_less_fee_scaled);
+		let input_amount_scaled = FeeRate::<PerMill>::from(T::BalanceToUnsignedInt::from(input_amount).into())
+			.checked_div(div_rate)
+			.ok_or::<&'static str>(Error::DivideByZero.into())?;
 
-		let denominator: u128 = (input_amount + U256::from(input_reserve))
```

Comment:
This would overflow previously as we try to convert `denominator` back into `u128`. Your changes keep the high precision up until the final price calculation which I think is a good improvement 🥇